gcloud-java-storage/src/main/java/com/google/gcloud/storage/Blob.java

```diff
@@ -0,0 +1,239 @@
+/**
+ * A Google cloud storage object.
+ */
+public final class Blob {
+
+  private final Storage storage;
+  private BlobInfo info;
+
+  public static class BlobSourceOption extends Option {
+
+    private static final long serialVersionUID = 214616862061934846L;
+
+    private BlobSourceOption(StorageRpc.Option rpcOption) {
+      super(rpcOption, null);
+    }
+
+    private Storage.BlobSourceOption convert(BlobInfo blobInfo) {
+      switch (rpcOption()) {
+        case IF_GENERATION_MATCH:
+          return Storage.BlobSourceOption.generationMatch(blobInfo.generation());
+        case IF_GENERATION_NOT_MATCH:
+          return Storage.BlobSourceOption.generationNotMatch(blobInfo.generation());
+        case IF_METAGENERATION_MATCH:
+          return Storage.BlobSourceOption.metagenerationMatch(blobInfo.metageneration());
+        case IF_METAGENERATION_NOT_MATCH:
+          return Storage.BlobSourceOption.metagenerationNotMatch(blobInfo.metageneration());
+        default:
+          throw new AssertionError("Unexpected enum value");
+      }
+    }
+
+    public static BlobSourceOption generationMatch() {
+      return new BlobSourceOption(StorageRpc.Option.IF_GENERATION_MATCH);
+    }
+
+    public static BlobSourceOption generationNotMatch() {
+      return new BlobSourceOption(StorageRpc.Option.IF_GENERATION_NOT_MATCH);
+    }
+
+    public static BlobSourceOption metagenerationMatch() {
+      return new BlobSourceOption(StorageRpc.Option.IF_METAGENERATION_MATCH);
+    }
+
+    public static BlobSourceOption metagenerationNotMatch() {
+      return new BlobSourceOption(StorageRpc.Option.IF_METAGENERATION_NOT_MATCH);
+    }
+
+    static Storage.BlobSourceOption[] convert(BlobInfo blobInfo, BlobSourceOption... options) {
+      Storage.BlobSourceOption[] convertedOptions = new Storage.BlobSourceOption[options.length];
+      int index = 0;
+      for (BlobSourceOption option : options) {
+        convertedOptions[index++] = option.convert(blobInfo);
+      }
+      return convertedOptions;
+    }
+  }
+
+  /**
+   * Construct a {@code Blob} object for the provided {@code BlobInfo}. The storage service is used
+   * to issue requests.
+   * 
+   * @param storage the storage service used for issuing requests
+   * @param info blob's info
+   */
+  public Blob(Storage storage, BlobInfo info) {
+    this.storage = checkNotNull(storage);
+    this.info = checkNotNull(info);
+  }
+
+  /**
+   * Construct a {@code Blob} object for the provided bucket and blob names. The storage service is
+   * used to issue requests.
+   * 
+   * @param storage the storage service used for issuing requests
+   * @param bucket bucket's name
+   * @param blob blob's name
+   */
+  public Blob(Storage storage, String bucket, String blob) {
+    this.storage = checkNotNull(storage);
+    this.info = BlobInfo.of(checkNotNull(bucket), checkNotNull(blob));
+  }
+
+  /**
+   * Return the blob's information.
+   */
+  public BlobInfo info() {
+    return info;
+  }
+
```

Comment:
> Having a Blob pinned to its info but using only its name and bucket to issue requests might be a bit counter-intuitive.

I am not sure about that. I mean, the user has that option by passing BlobSourceOption.generationMatch and/or the BlobSourceOptions.metagenerationMatch
options. Though clearly if we choose to go that route documentation should make it clear
that operations are not enforcing the version check by default.

> This sounds cool. But I would not put this as high priority task (nor include it in this PR) as optimization is not the main focus of the library. I would open an issue for this.

Yes,  I also didn't think we should do it in this PR but I do think having an optimized code (if it does not compromise the API) is something that we should care for. Also, for listing of many
files this could have a significant benefit.